plugins/hls-diagrams-plugin/src/Ide/Plugin/Diagrams.hs

```diff
@@ -0,0 +1,332 @@
+
+    -- Figure out what we hovered over
+    (IDE.HAR{hieAst, hieKind}, _posMap) <- getHieAst ide modulePathNorm
+
+    case hieKind of
+      IDE.HieFresh -> do
+        let ids = getIdsAtPoint hieAst _position
+        logWith recorder Info $ HoverOver ids
+        case ids of
+          [(name, typ)] | looksLikeDiagram typ -> do
+            tempDir  <- runAction ide $ IDE.useNoFile_ DiagramsTempDir
+            tempFile <- liftIO $ emptyTempFile tempDir "diagram.svg"
+            success  <- renderDiagram ide moduleInfo tempFile name
+            if success then do
+              logWith recorder Info (RenderedDiagram tempFile)
+              return $ Just LSP.Hover {
+                  _contents = LSP.HoverContents $ LSP.MarkupContent {
+                      _kind  = LSP.MkMarkdown
+                    , _value = Text.unlines [
+                            Text.pack $ "![](" ++ tempFile ++ ")"
+                          ]
+                    }
+                , _range  = Nothing -- Which range should be highlighted
+                }
+            else do
+              logWith recorder Info $ RenderingFailed
+              return Nothing
+          _otherwise ->
+            return Nothing
+      IDE.HieFromDisk{} ->
+        return Nothing
+  where
+    looksLikeDiagram :: IDE.Type -> Bool
+    looksLikeDiagram = ("Diagram" `isPrefixOf`) . IDE.printWithoutUniques
+
+{-------------------------------------------------------------------------------
+  Auxiliary: HLS
+-------------------------------------------------------------------------------}
+
+getIdsAtPoint :: IDE.HieASTs GHC.Type -> LSP.Position -> [(Name, IDE.Type)]
+getIdsAtPoint ast pos =
+     concatMap (mapMaybe (uncurry aux) . Map.toList . GHC.nodeIdentifiers) $
+       IDE.pointCommand ast pos GHC.nodeInfo
+  where
+    aux ::
+         IDE.Identifier
+      -> IDE.IdentifierDetails IDE.Type
+      -> Maybe (Name, IDE.Type)
+    aux (Left  _mod) _       = Nothing
+    aux (Right name) details = (name,) <$> GHC.identType details
+
+enableCodeGen :: MonadIO m => IdeState -> ModuleInfo -> m ()
+enableCodeGen ide ModuleInfo{modulePathNorm} = liftIO $ do
+    Eval.queueForEvaluation ide modulePathNorm
+    IDE.setSomethingModified
+      Shake.VFSUnmodified
+      ide
+      [Shake.toKey IDE.NeedsCompilation modulePathNorm]
+      "hls-diagrams-plugin"
+
+{-------------------------------------------------------------------------------
+  Auxiliary: GHC
+-------------------------------------------------------------------------------}
+
+-- | Unload any modules that are out of date
+unloadOutOfDate :: MonadIO m => IdeState -> HscEnv -> CatchErrors c m ()
+unloadOutOfDate ide env = runAction ide $ do
+    current <- IDE.currentLinkables
+    liftIO $ GHC.unload env $ map (uncurry aux) $ GHC.moduleEnvToList current
+  where
+    aux :: GHC.NoCompat.Module -> UTCTime -> GHC.Linkable
+    aux mod time = GHC.LM time mod []
+
+initialiseSessionForEval ::
```

Comment:
Can these bits be shared with the eval plugin? It looks like you already depend on it. GHC session management is fiddly at the best of times, it would be great if we can deduplicate it!

Reply:
I was going to ask the same thing. The Eval plugin already contains code to do interactive evaluation, so I think it would be best if hls-diagrams-plugin reused more of it!